File: js/131_PalindromePartitioning.js
/**
 * @param {string} s
 * @return {string[][]}
 */
var partition = function(s) {
    var ret = [];
    backtrace(ret,[],0,s);
    return ret;
    
    function isPalidrome(str,begin,end){
        while(begin < end){
            if(str[begin] !== str[end]) return false;
            begin++;
            end--;
        }
        return true;
    }
    
    function backtrace(ret, path, index, s){
        if(index === s.length){
            ret.push(path.concat());
            return;
        }
        for(var i = index; i < s.length; ++i){
            if(isPalidrome(s,index,i)){
                path.push(s.substring(index,i+1));
                backtrace(ret,path,i+1,s);
                path.pop();
            }
        }
    }
};

File: cpp/131_PalindromePartitioning.cpp
class Solution {
public:
    vector<vector<string>> partition(string s) {
        vector<vector<string>> ret;
        vector<string> path;
        backtrace(ret,path,0,s);
        return ret;
    }
    
private:
    void backtrace(vector<vector<string>>& ret, vector<string>& path, int index, string& s){
        if(index == s.length()){
            ret.push_back(path);
            return;
        }
        for(int i = index; i < s.length(); ++i){
            if(isPalindrome(s,index,i)){
                path.push_back(s.substr(index,i-index+1));
                backtrace(ret,path,i+1,s);
                path.pop_back();
            }
        }
    }
    
    bool isPalindrome(string s, int begin, int end){
        while(begin < end){
            if(s[begin] != s[end]) return false;
            begin++;
            end--;
        }
        return true;
    }
};